otel/otel.go

```diff
@@ -0,0 +1,171 @@
+}
+
+func New(ctx context.Context, cfg Config) (shutdown func(context.Context) error, err error) {
+	var shutdownFuncs []func(context.Context) error
+
+	// shutdown calls cleanup functions registered via shutdownFuncs.
+	// The errors from the calls are joined.
+	// Each registered cleanup will be invoked once.
+	shutdown = func(ctx context.Context) error {
+		var err error
+		for _, fn := range shutdownFuncs {
+			err = errors.Join(err, fn(ctx))
+		}
+		shutdownFuncs = nil
+		return err
+	}
+
+	// handleErr calls shutdown for cleanup and makes sure that all errors are returned.
+	handleErr := func(inErr error) {
+		err = errors.Join(inErr, shutdown(ctx))
+	}
+
+	// Set up meter provider.
+	meterProvider, err := newMeterProvider(ctx, cfg)
+	if err != nil {
+		handleErr(err)
+		return
+	}
+	shutdownFuncs = append(shutdownFuncs, meterProvider.Shutdown)
+	otelsdk.SetMeterProvider(meterProvider)
+
+	globalWrapper = &OtelWrapper{
+		meterProvider: meterProvider,
+		globalMeter:   meterProvider.Meter(cfg.ServiceName),
+		config:        cfg,
+	}
+
+	return
+}
+
+func newMeterProvider(ctx context.Context, cfg Config) (*sdkmetric.MeterProvider, error) {
+	if cfg.ServiceName == "" {
+		return nil, errors.New("ServiceName is required")
+	}
+	if cfg.MetricsExporter == nil {
+		exporter, err := newMetricsExporter(ctx, cfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load exporter", err)
+		}
+		cfg.MetricsExporter = exporter
+	}
+	if cfg.MetricsReader == nil {
+		cfg.MetricsReader = sdkmetric.NewPeriodicReader(cfg.MetricsExporter, sdkmetric.WithInterval(cfg.CollectionInterval))
+	}
+
+	res, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(cfg.ServiceName),
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create resource: %w", err)
+	}
+
+	// Initialize MeterProvider
+	provider := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(cfg.MetricsReader),
+		sdkmetric.WithResource(res),
+	)
+
+	return provider, nil
+}
+
+func newMetricsExporter(ctx context.Context, cfg Config) (sdkmetric.Exporter, error) {
+	if cfg.Debug {
+		return stdoutmetric.New(stdoutmetric.WithPrettyPrint())
+	}
+	switch cfg.ExporterType {
+	case "http":
+		return otlpmetrichttp.New(ctx)
+	case "grpc":
+		return otlpmetricgrpc.New(ctx)
+	default:
+		return nil, errors.New("invalid exporter type")
+	}
+}
+
+// GetGlobalMeter returns the global Meter
+func GetGlobalMeter() metric.Meter {
+	if globalWrapper == nil {
+		panic("Global OtelWrapper is not initialized")
+	}
+	return globalWrapper.globalMeter
+}
+
+// SendMetric wraps the meter to send a metric
+func SendMetric(ctx context.Context, name string, value float64, description string) error {
+	counter, err := GetGlobalMeter().Float64Counter(
+		name,
+		metric.WithDescription(description),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create counter: %w", err)
+	}
+	counter.Add(ctx, value)
+	return nil
+}
```

Comment:
Same for that, i would handle this in a separate PR as there's multiple issue. Step by step.